app/models/transformation_mapping/vm_migration_validator.rb

```diff
@@ -0,0 +1,144 @@
+  end
+
+  def select_vms
+    valid_list = []
+
+    @mapping.transformation_mapping_items.where(:source_type => EmsCluster).collect(&:source).collect(&:vms).flatten.each do |vm|
+      reason = validate_vm(vm, true)
+      valid_list << VmMigrateStruct.new(vm.name, vm, VM_VALID, reason) if reason == VM_VALID
+    end
+
+    {"valid" => valid_list}
+  end
+
+  def identify_vms
+    valid_list = []
+    invalid_list = []
+    conflict_list = []
+
+    @vm_list.each do |row|
+      vm_name = row['name']
+
+      if vm_name.blank?
+        invalid_list << VmMigrateStruct.new('', nil, VM_INVALID, VM_EMPTY_NAME)
+        next
+      end
+
+      query = Vm.where(:name => vm_name)
+      query = query.where(:uid_ems => row['uid_ems']) if row['uid_ems'].present?
+      query = query.joins(:host).where(:hosts => {:name => row['host']}) if row['host'].present?
+      query = query.joins(:ext_management_system).where(:ext_management_systems => {:name => row['provider']}) if row['provider'].present?
+
+      vms = query.to_a
+      if vms.size.zero?
+        invalid_list << VmMigrateStruct.new(vm_name, nil, VM_INVALID, VM_NOT_EXIST)
+      elsif vms.size == 1
+        vm = vms.first
+        reason = validate_vm(vm, false)
+        if reason == VM_VALID
+          valid_list << VmMigrateStruct.new(vm.name, vm, VM_VALID, reason)
+        else
+          invalid_list << VmMigrateStruct.new(vm.name, vm, VM_INVALID, reason)
+        end
+      else
+        vms.each { |v| conflict_list << VmMigrateStruct.new(v.name, v, VM_CONFLICT, VM_CONFLICT) }
+      end
+    end
+
+    {
+      "valid"      => valid_list,
+      "invalid"    => invalid_list,
+      "conflicted" => conflict_list
+    }
+  end
+
+  def validate_vm(vm, quick = true)
+    # a valid vm must find all resources in the mapping and has never been migrated
+    invalid_list = []
+
+    unless valid_cluster?(vm)
+      invalid_list << "cluster: %{name}" % {:name => vm.ems_cluster.name}
+      return no_mapping_msg(invalid_list) if quick
+    end
+
+    invalid_storages = unmapped_storages(vm)
+    if invalid_storages.present?
+      invalid_list << "storages: %{list}" % {:list => invalid_storages.collect(&:name).join(", ")}
+      return no_mapping_msg(invalid_list) if quick
+    end
+
+    invalid_lans = unmapped_lans(vm)
+    if invalid_lans.present?
+      invalid_list << "lans: %{list}" % {:list => invalid_lans.collect(&:name).join(', ')}
+      return no_mapping_msg(invalid_list) if quick
+    end
+
+    return no_mapping_msg(invalid_list) if invalid_list.present?
+    vm_migration_status(vm)
+  end
+
+  def vm_migration_status(vm)
+    vm_as_resources = ServiceResource.where(:resource => vm).includes(:service_template).where(:service_templates => {:type => "ServiceTemplateTransformationPlan"})
+
+    # VM has not been migrated before
+    return VM_VALID if vm_as_resources.blank?
+
+    return VM_MIGRATED if vm_as_resources.any? { |rsc| rsc.status == ServiceResource::STATUS_COMPLETED }
```

Comment:
We were trying to query the DB once, then process the results from the in memory objects.

Reply:
This feels like a thing that would be better to do in the database rather than in ruby.

The `.blank?` above could be `.empty?` which will select count instead of loading all of the records.
Then this could be another `.where(:status => ... ).empty?` and still avoid loading all of those objects.